midend/flattenHeaders.h

```diff
@@ -0,0 +1,187 @@
+/*
+    Copyright 2018 MNK Consulting, LLC.
+    http://mnkcg.com
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _MIDEND_FLATTENHEADERS_H_
+#define _MIDEND_FLATTENHEADERS_H_
+
+#include "ir/ir.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+
+namespace P4 {
+
+struct HeaderTypeReplacement : public IHasDbPrint {
+    HeaderTypeReplacement(const P4::TypeMap* typeMap, const IR::Type_Header* type);
+
+    // Maps nested field names to final field names.
+    // In our example this could be:
+    // .t.s.a -> _t_s_a0;
+    // .t.s.b -> _t_s_b1;
+    // .t.y -> _t_y2;
+    // .x -> _x3;
+    std::map<cstring, cstring> fieldNameRemap;
+    // Holds a new flat type
+    // struct M {
+    //    bit _t_s_a0;
+    //    bool _t_s_b1;
+    //    bit<6> _t_y2;
+    //    bit<3> _x3;
+    // }
+    const IR::Type* replacementType;
+    virtual void dbprint(std::ostream& out) const {
+        out << replacementType;
+    }
+
+    // Helper for constructor
+    void flatten(const P4::TypeMap* typeMap,
+                 cstring prefix,
+                 const IR::Type* type,
+                 IR::IndexedVector<IR::StructField> *fields);
+};
+
+/**
+ * Maintains a map between an original header type and its
+ * replacement.
+ */
+struct NestedHeaderMap {
+    P4::ReferenceMap* refMap;
+    P4::TypeMap* typeMap;
+
+    ordered_map<cstring, HeaderTypeReplacement*> replacement;
+
+    NestedHeaderMap(P4::ReferenceMap* refMap, P4::TypeMap* typeMap):
+            refMap(refMap), typeMap(typeMap)
+    { CHECK_NULL(refMap); CHECK_NULL(typeMap); }
+    void createReplacement(const IR::Type_Header* type);
+    HeaderTypeReplacement* getReplacement(const cstring name) const
+    { return ::get(replacement, name); }
+    bool empty() const { return replacement.empty(); }
+};
+
+/**
+Find the types to replace and insert them in the nested struct map.
+ */
+class FindHeaderTypesToReplace : public Inspector {
+    NestedHeaderMap* map;
+ public:
+    explicit FindHeaderTypesToReplace(NestedHeaderMap* map): map(map) {
+        setName("FindHeaderTypesToReplace");
+        CHECK_NULL(map);
+    }
+    bool preorder(const IR::Type_Header* type) override;
+};
+
+/**
+ * This pass flattens any nested struct inside a P4 header
+
+ * Should be run before the flattenInterfaceStructs pass.
+
+For example, see the data structures below.
+
+struct alt_t {
+  bit<1> valid;
+  bit<7> port;
+};
+
+struct row_t {
```

Comment:
How does this pass deal with annotations on the `struct`? For instance, suppose there is an annotation `@custom_anno` on `row_t`, can you ensure the annotation is not lost?

Reply:
Both midend passes for flattening, do not mess up any annotation.  I will add an annotation to the P4 program and checkin the test in the next commit.  This is what I did:

@MNK_annotation("(test flatten)")
struct row_t {
  alt_t alt0;
  alt_t alt1;
};

See my test log below.  In the test I have enabled both FlattenHeaders and FlattenInterfaceStructs passes and dumped the MidEndLast IR.  The IR shows the annotation is not lost.

hemant@ubuntu:~/p4c-1/build$ ./p4c-bm2-ss --std p4-16 ../testdata/p4_16_samples/issue383-bmv2.p4 --top4 MidEndLast --dump ~
hemant@ubuntu:~/p4c-1/build$ 

A portion of the IR dump is included below.

#include <core.p4>
#include <v1model.p4>

struct alt_t {
    bit<1> valid;
    bit<7> port;
}

@MNK_annotation("(test flatten)") struct row_t {
    alt_t alt0;
    alt_t alt1;
}

Reply:
@hanw: I think the PSA implementation in p4c is not complete and thus we have not tested this PR with the bmv2 PSA switch.  If you have a more complete private PSA implementation, please take p4c from this branch and test your code.  We have uses bmv2 simple switch to test the flattening passes. 

Also, when you have a p4c backend that supports nested structs of this PR, please test your backend as well.  In that test, edit midend.cpp for the platform to not include the two midend passes in FlattenHeaders and FlattenInterfaceStructs.  Maybe even the NestedStructs pass - this pass flattens local variables.   

thanks!